pkg/utils/metadata.go

```diff
@@ -75,3 +77,36 @@ func ClusterName(clusterName string, client metadataClient.Client) (string, erro
 	}
 	return clusterName, nil
 }
+
+// ZoneToRegion converts a GKE zone to its region
+func ZoneToRegion(zone string) (string, error) {
+	fields := strings.Split(zone, "-")
+	if len(fields) == 3 {
+		return fmt.Sprintf("%s-%s", fields[0], fields[1]), nil
+	}
+	// We can as well treat xx-xx as region and simply return,
+	// but let's be strict here
+	return "", fmt.Errorf("zone %s is not valid", zone)
+}
+
+// ClusterRegion returns the region of the cluster
+func ClusterRegion(clusterRegion string, clientCreator func() metadataClient.Client) (string, error) {
+	if clusterRegion != "" {
+		return clusterRegion, nil
+	}
+	zone, err := clientCreator().Zone()
+	if err != nil {
+		return "", err
+	}
+	return ZoneToRegion(zone)
+}
+
+// ClusterRegionGetter is a handler that gets cluster region
+type ClusterRegionGetter func() (string, error)
+
+// NewClusterRegionGetter returns CluterRegionGetter in production code
+func NewClusterRegionGetter() ClusterRegionGetter {
+	return func() (string, error) {
+		return ClusterRegion("", defaultMetadataClientCreator)
```

Comment:
This isn't expected to change for the lifetime of the process, so I recommend using a `sync.Once` to get the value once, cache it, and return that value from then on without calling the metadata server.

Reply:
sync.Once might not work due to possible error condition, how about the new caching way?